emul: espi: Add ACPI Shared Memory functions

yperess opened the pull request:
Add the bare minimum to set and access the ACPI shared memory via the
eSPI emulator.

Signed-off-by: Yuval Peress <peress@chromium.org>

yperess commented:
@semihalf-niedzwiecki-dawid could you take a look and let me know if this seems reasonable to you?

semihalf-niedzwiecki-dawid commented on drivers/espi/espi_emul.c:
```diff
@@ -29,6 +29,10 @@ struct espi_emul_data {
 	sys_slist_t callbacks;
+#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
+	/** ACPI Shared memory. */
+	uint8_t shm_acpi_mmap[CONFIG_ESPI_EMUL_HOST_ACPI_SHM_REGION_SIZE];
```
I think the mmap should be stored on the host side.

yperess replied:
Done

semihalf-niedzwiecki-dawid commented on drivers/espi/espi_emul.c:
```diff
@@ -87,6 +91,44 @@ static bool espi_emul_get_channel_status(const struct device *dev, enum espi_cha
+	switch (op) {
+#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
+	case EACPI_GET_SHARED_MEMORY:
+		*data = (uint32_t)dev_data->shm_acpi_mmap;
```
Here, we should send a request for the pointer of the mmap to the host via lpc_reqest, as it is done in espi_emul_receive_vwire.

yperess replied:
Done.

semihalf-niedzwiecki-dawid commented on drivers/espi/espi_emul.c:
```diff
@@ -172,12 +214,17 @@ static struct emul_espi_driver_api emul_espi_driver_api = {
 	.trigger_event = emul_espi_trigger_event,
 	.find_emul = espi_emul_find,
+#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
+	.set_acpi_shm = espi_emul_set_acpi_shm,
```
This one would be on the host side.

yperess replied:
Done.

semihalf-niedzwiecki-dawid commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -217,6 +217,19 @@ int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data)
+int emul_espi_host_set_acpi_shm(const struct device *espi_dev, uint8_t *data,
+				size_t size)
+{
+	struct emul_espi_driver_api *api =
```
Here we would set the mmap memory stored in the host data.

yperess replied:
Done.

semihalf-niedzwiecki-dawid commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -217,6 +217,19 @@ int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data)
+
 /* Device instantiation */
 static struct emul_espi_device_api ap_emul_api = {
 	.set_vw = emul_host_set_vw,
```
This api should be extended to handle the lpc_request and return the mmap pointer.

yperess replied:
Done.

semihalf-niedzwiecki-dawid commented on include/drivers/espi_emul.h:
```diff
@@ -88,11 +99,13 @@ typedef struct espi_emul *(*emul_find_emul)(const struct device *dev,
 	emul_espi_api_set_vw set_vw;
 	emul_espi_api_get_vw get_vw;
+#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
+	emul_espi_api_get_acpi_shm get_acpi_shm;
```
Can we add write_lpc_request instead and return a proper value on the host side based on the opcode? Or do you think the current approach is better?

yperess replied:
Just to make sure I'm on the same page. It looks like the implementation of `read_lpc_request` for the NPCX chip is to basically just return the address of the shared memory. I thought about implementing that and basically only supporting the one op code for now (it would look similar to `host_subs_npcx.c::npcx_host_periph_read_request()` and handle `EACPI_GET_SHARED_MEMORY` only).

The `write_lpc_request` doesn't seem to access the shared memory (AFAICT), is there something I might be missing in how the emulator should handle the shared memory?

semihalf-niedzwiecki-dawid replied:
You are right, it should be `read_lpc_request`, similar to the `host_subs_npcx.c::npcx_host_periph_read_request`.

yperess replied:
I'm looking at this some more and I think I might have confused myself :)

My understanding of the PR code as it is currently is that:
* I added a `.get_acpi_shm` function to the emulated eSPI host device. This is done in `emul_espi_host.c`.
* I then set both `.read_lpc_request` and `.write_lpc_request` for the emulated eSPI device in `espi_emul.c`. This means that the emulated eSPI device should now be able to call those functions: `espi_read_lpc_request` and `espi_write_lpc_request`.
* The implementation of the write function is a stub and will always return an error for now
* The implementation of the read will only handle the `EACPI_GET_SHARED_MEMORY` operation and will call the host function to directly get the shared memory address and cast it to a `uint32_t`.
* During the test, we can mock host interaction with the memory by also calling the same function to get the memory address and read/write to it.

Adding `read_lpc_request` to the host emulator might not make sense for all op codes (I'm not very familiar with eSPI, but that's my understanding). So instead I think that for the host emulator to have direct access functions to the specific components such as the shared memory makes the most sense.

One thing that I do think might come in handy (but I didn't add yet until we find a use for it) is to add a host API to get the shared memory directly. This would just replicate the typedef for `emul_espi_api_get_acpi_shm` in `espi_emul.h` but would allow global access such as `intptr_t shm = emul_espi_host_get_acpi_shm(espi_host);`.

semihalf-niedzwiecki-dawid replied:
Regarding `emul_espi_host_get_acpi_shm`, I was about to suggest you to add this kind of function. I think it is better to have  access to the share memory via the emulated host, instead of calling functions from real eSPI API(which is of course also emulated).

I'm also not an eSPI expert, but dedicated host functions sounds good for me.

Regarding the bullets, my understanding of the PR is the same, except for the last point once the `emul_espi_host_get_acpi_shm` function is added.

yperess replied:
Sounds good, I'll make the changes soon. I also just thought of a good unit test that I think would work for this so I'll bundle that up in the same PR.

Commit:
emul: espi: Add ACPI Shared Memory functions

Add the bare minimum to set and access the ACPI shared memory via the
eSPI emulator.

Signed-off-by: Yuval Peress <peress@chromium.org>

yperess commented:
I've added some test cases and rebased this on top of #37192

semihalf-niedzwiecki-dawid commented on include/drivers/espi_emul.h:
```diff
@@ -152,6 +165,18 @@ int emul_espi_host_send_vw(const struct device *espi_dev,
+ * Get the host device's ACPI shared memory start address. The size of the region is
+ * CONFIG_EMUL_ESPI_HOST_ACPI_SHM_REGION_SIZE.
+ *
+ * @param espi_dev eSPI host emulator.
```
The other similar functions(`emul_espi_host_send_vw` and `emul_espi_host_port80_write`) take a pointer to the eSPI emulation controller not host. I think we should keep this consistent.

yperess replied:
Done, sorry for the confusion. I thought we wanted to be able to get the address via the emulator directly.

semihalf-niedzwiecki-dawid replied:
Please, change the comment to eSPI controller emulator.

semihalf-niedzwiecki-dawid commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -217,10 +221,29 @@ int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data)
 }
 
+#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
+uintptr_t emul_espi_dev_get_acpi_shm(struct espi_emul *emul)
```
static

semihalf-niedzwiecki-dawid commented on subsys/emul/espi/emul_espi_host.c:
```diff
@@ -217,10 +221,29 @@ int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data)
+	return (uintptr_t)data->shm_acpi_mmap;
+}
+
+uintptr_t emul_espi_host_get_acpi_shm(const struct emul *emulator)
```
As I commented above.

yperess commented:
I've update the comment and rebased again off the same PR #37192. Note that this will not build until that pull request is merged.

semihalf-niedzwiecki-dawid commented:
Looks good now.

yperess commented:
I gave up basing this ontop of the test suite PR (it's taking way too long to merge). Added `test_main` in the tests for now.

albertofloyd commented on drivers/espi/espi_emul.c:
```diff
@@ -87,6 +87,49 @@ static bool espi_emul_get_channel_status(const struct device *dev, enum espi_cha
+	struct espi_emul_data *emul_data = dev->data;
+
+	if (!(emul_data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
+		LOG_ERR("bad channel vwire");
```
ENOTSUP or EINVAL goes more accordingly with eSPI driver API

yperess commented:
Ping, this is blocking some work on Chromium's test efforts.

yperess commented:
@sjg20 Would you be able to take a look at this PR? We need for some of our tests.
